File: LogisticRegression.py
import numpy as np

class LogisticRegression:
    def __init__(self, learning_rate=0.01, num_iterations=500):
        self.learning_rate = learning_rate
        self.num_iterations = num_iterations
        self.weights = None
        self.bias = None

    def fit(self, X, y):
        num_samples, num_features = X.shape
       
        # Initialize weights and bias
        self.weights = np.zeros(num_features)
        self.bias = 0

        # Gradient descent
        for _ in range(self.num_iterations):
            linear_model = np.dot(X, self.weights) + self.bias
            y_pred = self._sigmoid(linear_model)

            # Update weights and bias
            dw = (1 / num_samples) * np.dot(X.T, (y_pred - y))
            db = (1 / num_samples) * np.sum(y_pred - y)
            print("dw: ", dw)
            print("db: ", db)
            
            self.weights -= self.learning_rate * dw
            self.bias -= self.learning_rate * db

    def predict(self, X):
        linear_model = np.dot(X, self.weights) + self.bias
        y_pred = self._sigmoid(linear_model)
        y_pred_class = [1 if pred >= 0.5 else 0 for pred in y_pred]
        return y_pred_class

    def _sigmoid(self, x):
        return 1 / (1 + np.exp(-x))
    
    
    
if __name__ == '__main__':
    # Imports
    import numpy as np
    from sklearn.model_selection import train_test_split
    from sklearn import datasets
    import matplotlib.pyplot as plt

    # Load dataset
    bc = datasets.load_breast_cancer()
    X, y = bc.data, bc.target
    
    # Split dataset
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1234)

    # Initialize model
    model = LogisticRegression(learning_rate=0.0001, num_iterations=500)

    # Train model
    model.fit(X_train, y_train)

    # Predict
    predictions = model.predict(X_test)

    # Evaluate
    accuracy = np.sum(predictions == y_test) / len(y_test)
    print(accuracy)

File: LogisticRegression.js
class LogisticRegression {
  constructor(learningRate = 0.1, numIterations = 100) {
    this.learningRate = learningRate;
    this.numIterations = numIterations;
    this.weights = [];
    this.bias = 0;
  }

  train(features, labels) {
    // Initialize weights and bias
    this.weights = new Array(features[0].length).fill(0);

    // Gradient descent
    for (let i = 0; i < this.numIterations; i++) {
      const scores = this.predict(features);
      const gradients = this.computeGradients(features, labels, scores);

      // Update weights and bias
      for (let j = 0; j < this.weights.length; j++) {
        this.weights[j] -= this.learningRate * gradients[j];
      }
      this.bias -= this.learningRate * gradients[this.weights.length];
    }
  }

  predict(features) {//output i*1
    const scores = [];

    for (let i = 0; i < features.length; i++) {
      let score = this.bias;

      for (let j = 0; j < this.weights.length; j++) {
        score += features[i][j] * this.weights[j];
      } 

      scores.push(this.sigmoid(score));
    }

    return scores;
  }

  computeGradients(features, labels, scores) {
    const gradients = new Array(this.weights.length + 1).fill(0);

    for (let i = 0; i < features.length; i++) {
      const error = labels[i] - scores[i];

      for (let j = 0; j < this.weights.length; j++) {
        gradients[j] += error * features[i][j];
      }

      gradients[this.weights.length] += error;
    }

    return gradients;
  }

  sigmoid(x) {
    return 1 / (1 + Math.exp(-x));
  }
}
